### PR TITLE
propose pre-transfer leader

### PR DESCRIPTION
**Pull request by BusyJay**
The proposal aims to make raft leadership transferring more smooth.

### CODE REVIEW

File: text/2020-01-20-pre-transfer-leader.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+## Detailed design
+
+To minimize the probability of failure, a leader can send a pre-request to ask
+target follower whether it's suitable to become a leader. The follower can
+check all the situations that can block it from becoming a good leader. If all
+checks pass, follower replies to leader an ack to start actual leadership
+transferring.
+
+To avoid stale ack, pre-request should be attached with leader's current commit
```
Comment (hicqu) [inline text/2020-01-20-pre-transfer-leader.md]: Do we need a mechanism for pre transfer timeout?

Comment (BusyJay) [reply text/2020-01-20-pre-transfer-leader.md]: I don't think so. The client, pd for example, should retry by itself.

File: text/2020-01-20-pre-transfer-leader.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+election because of unapplied conf change or message loss. Even it starts
+election and becomes leader at the end, it may not start to respond to requests
+because logs from last term is still being committed and applied. In the
+practice of TiKV, we observe a slow new leader can take seconds or minutes to
+be active, which is a serious problem needs to be solved.
+
+## Detailed design
+
+To minimize the probability of failure, a leader can send a pre-request to ask
+target follower whether it's suitable to become a leader. The follower can
```
Comment (Hoverbear) [inline text/2020-01-20-pre-transfer-leader.md]:
```suggestion
target follower whether it's suitable to become a leader. The follower must
```

**Review by hicqu — APPROVED** [review]

**Comment by zhangjinpeng1987** [comment]
@BusyJay Please fix the ci.

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Pull request merged**